aospy_synthetic/db/sqlalchemy/sqlalchemy_config.py

```diff
@@ -0,0 +1,422 @@
+
+    First we check to see if the object has been added to the current session;
+    then we check to see if the object is already in the database.  If neither
+    of those are true, the object is added to the database.
+
+    Parameters
+    ----------
+    session : Session
+        Active sqlalchemy Session.
+    cls
+        Database object class associated with the provided aospy core object.
+    queryfunc : function
+        Function to call to return a sqlalchemy Query object associated with
+        the aospy core object.
+    constructor : function
+        Constructor of database row object.
+    AospyObj
+        Aospy core object to insert into the database.
+
+    Returns
+    -------
+    DBObj
+        Database row object.
+    """
+    cache = getattr(session, '_unique_cache', None)
+    if cache is None:
+        session._unique_cache = cache = {}
+
+    key = hash(AospyObj)
+
+    # First check to see if the row was added to the session
+    if key in cache:
+        _set_metadata_attrs(cache[key], AospyObj)
+        return cache[key]
+
+    # Then check if row is already in the DB
+    else:
+        with session.no_autoflush:
+            q = session.query(cls)
+            q = queryfunc(q, AospyObj)
+            obj = q.first()
+
+            # If it is not in the DB or session cache, create a new row
+            if not obj:
+                obj = constructor(session, AospyObj)
+            else:
+                _set_metadata_attrs(obj, AospyObj)
+        cache[key] = obj
+    return obj
+
+
+class UniqueMixin(object):
+    """Superclass for all sqlalchemy row objects for use with aospy to
+    enforce that no duplicate entries will be saved in the database.
+
+    Uniqueness checking using this superclass requires that all aospy
+    core objects implement a hash function, which can be used to
+    distinguish between distinct aospy core objects.
+
+    This recipe was adapted from `SQLAlchemy's Bitbucket
+    Repo<https://bitbucket.org/zzzeek/sqlalchemy/wiki/
+    UsageRecipes/UniqueObject>`.
+    """
+
+    # Maps a non DB object attribute name in the DB object to
+    # an attribute name in the corresponding aospy object
+    _metadata_attrs = {}
+
+    # Maps a DB object attribute name in the DB object to
+    # another dictionary, which maps 'db_cls' to the DB object
+    # class associated with the type of the DB object the attribute
+    # points to and maps 'aospy_obj_attr' to the attribute name in the
+    # aospy object associated with the DB object
+    _db_attrs = {}
+    hash = Column(String)
+
+    @staticmethod
+    def unique_filter(query, AospyObj):
+        """Returns a database filter object for a given aospy core object.
+
+        Paramters
+        ---------
+        query : Query
+            Sqlalchemy Query object associated with a given database
+            row object.
+        AospyObj
+            Aospy core object.
+        """
+        return query.filter_by(hash=hash(AospyObj))
```

Comment:
Good catch -- this made me nervous for a second.  I did a quick test, and within this particular scope ```hash``` is still the built-in hash function (and not what is defined in L120), but I agree this is *very* confusing.  I have renamed the column in the DB objects to ```hashcode```.